File: connectors/peertube-ingest/app/peertube_client.py
"""Uploads to the PeerTube API (`POST /api/v1/videos/upload`), embedding meeting
metadata in the title/description (study 2.12 line 589). Async via `httpx.AsyncClient`.
"""

from __future__ import annotations

import os
from typing import Dict, Optional

import httpx

from .types import MeetingMetadata

PEERTUBE_BASE_URL = os.environ.get("PEERTUBE_BASE_URL", "https://tube.example.org")
PEERTUBE_ACCESS_TOKEN = os.environ.get("PEERTUBE_ACCESS_TOKEN", "")
PEERTUBE_CHANNEL_ID = os.environ.get("PEERTUBE_CHANNEL_ID", "1")
# Default visibility of uploaded recordings (study 2.12: "per-video visibility
# management -- private, internal, unlisted"). 2 = "internal" on the PeerTube
# API side (visible only to users logged into the instance).
PEERTUBE_DEFAULT_PRIVACY = int(os.environ.get("PEERTUBE_DEFAULT_PRIVACY", "2"))


async def upload_to_peertube(
    object_key: str,
    metadata: MeetingMetadata,
    file_stream: bytes,
    file_size_bytes: int,
    http_client: Optional[httpx.AsyncClient] = None,
) -> Dict[str, str]:
    title = f"{metadata.title} ({metadata.date})" if metadata.date else metadata.title
    description = (
        f"Participants: {', '.join(metadata.participants)}\nSource: {object_key}"
        if metadata.participants
        else f"Source: {object_key}"
    )

    file_name = object_key.split("/")[-1] if object_key else "recording.mp4"
    data = {
        "channelId": PEERTUBE_CHANNEL_ID,
        "name": title[:120],
        "description": description,
        "privacy": str(PEERTUBE_DEFAULT_PRIVACY),
    }
    files = {"videofile": (file_name, file_stream)}
    headers = {"Authorization": f"Bearer {PEERTUBE_ACCESS_TOKEN}"}
    url = f"{PEERTUBE_BASE_URL}/api/v1/videos/upload"

    owns_client = http_client is None
    client = http_client or httpx.AsyncClient()
    try:
        response = await client.post(url, data=data, files=files, headers=headers)
    finally:
        if owns_client:
            await client.aclose()

    if response.status_code >= 400:
        raise RuntimeError(f"PeerTube upload failed with status {response.status_code}")

    payload = response.json()
    video = payload.get("video") or {}
    video_id = video.get("uuid") or str(video.get("id") or "")
    return {"peertube_video_id": video_id}

File: connectors/peertube-ingest/src/peertube-client.ts
import type { Readable } from "node:stream";
import { MeetingMetadata } from "./types";

const PEERTUBE_BASE_URL = process.env.PEERTUBE_BASE_URL ?? "https://tube.example.org";
const PEERTUBE_ACCESS_TOKEN = process.env.PEERTUBE_ACCESS_TOKEN ?? "";
const PEERTUBE_CHANNEL_ID = process.env.PEERTUBE_CHANNEL_ID ?? "1";
// Default visibility of uploaded recordings (study 2.12: "per-video visibility
// management -- private, internal, unlisted"). 2 = "internal" on the PeerTube
// API side (visible only to users logged into the instance).
const PEERTUBE_DEFAULT_PRIVACY = Number(process.env.PEERTUBE_DEFAULT_PRIVACY ?? 2);

export interface UploadOptions {
  objectKey: string;
  metadata: MeetingMetadata;
  fileStream: Readable | Buffer;
  fileSizeBytes: number;
}

export interface UploadResult {
  peertubeVideoId: string;
}

/**
 * Uploads to the PeerTube API (`POST /api/v1/videos/upload`), embedding
 * meeting metadata in the title/description (study 2.12 line 589).
 * The PeerTube API expects `multipart/form-data`; built here by hand via
 * `FormData`/`Blob` (natively available since Node 18+) rather than an
 * extra dependency.
 */
export async function uploadToPeerTube(
  options: UploadOptions,
  fetchImpl: typeof fetch = fetch
): Promise<UploadResult> {
  const { metadata, objectKey } = options;
  const title = metadata.date ? `${metadata.title} (${metadata.date})` : metadata.title;
  const description = metadata.participants.length
    ? `Participants: ${metadata.participants.join(", ")}\nSource: ${objectKey}`
    : `Source: ${objectKey}`;

  const form = new FormData();
  form.set("channelId", PEERTUBE_CHANNEL_ID);
  form.set("name", title.slice(0, 120));
  form.set("description", description);
  form.set("privacy", String(PEERTUBE_DEFAULT_PRIVACY));
  const body = Buffer.isBuffer(options.fileStream)
    ? options.fileStream
    : await streamToBuffer(options.fileStream);
  form.set("videofile", new Blob([body]), objectKey.split("/").pop() ?? "recording.mp4");

  const response = await fetchImpl(`${PEERTUBE_BASE_URL}/api/v1/videos/upload`, {
    method: "POST",
    headers: { Authorization: `Bearer ${PEERTUBE_ACCESS_TOKEN}` },
    body: form,
  });

  if (!response.ok) {
    throw new Error(`PeerTube upload failed with status ${response.status}`);
  }

  const data = (await response.json()) as { video?: { id?: number | string; uuid?: string } };
  const id = data.video?.uuid ?? String(data.video?.id ?? "");
  return { peertubeVideoId: id };
}

async function streamToBuffer(stream: Readable): Promise<Buffer> {
  const chunks: Buffer[] = [];
  for await (const chunk of stream) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  }
  return Buffer.concat(chunks);
}
